### PR TITLE
Feature/subscriptions

### PR DESCRIPTION
**Pull request by tokland**
Add subscribe/unsubscribe actions to Chart/EventChart/EventReport/Map/ReportTable. The implementation and endpoints mimic the _favorite_ field:

_POST | DELETE_ _/api/OBJECT_TYPE/ID/subscriber_

This PR aims to provide the backend infrastructure for subscriptions. We will be creating other PRs in the coming weeks that will use this functionality to implement notifications, weekly digest, etc"

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 1]
```diff
@@ -1214,4 +1217,50 @@ public void setParentGraphMap( Map<String, String> parentGraphMap )
...
+    @JacksonXmlProperty( namespace = DxfNamespaces.DXF_2_0 )
+    public boolean isSubscribed()
+    {
+        User user = UserContext.getUser();
+
+        return user != null && subscribers != null ? subscribers.contains( user.getUid() ) : false;
+    }
+
+    @Override
+    public boolean subscribe( User user )
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java]: Would it make sense to just require the userUid here? there are cases where we already have the user uid available, but not the full object.. with this code we would have to do userService.get(...) and then send this object... same applies to unsubscribe

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java [hunk 2]
```diff
@@ -1214,4 +1217,50 @@ public void setParentGraphMap( Map<String, String> parentGraphMap )
...
+
+    public void setSubscribers( Set<String> subscribers )
+    {
+        this.subscribers = subscribers;
+    }
+
+    @Override
+    @JsonProperty
+    @JacksonXmlProperty( namespace = DxfNamespaces.DXF_2_0 )
+    public boolean isSubscribed()
```
Comment (mortenoh) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java]: This doesn't really feel right.. is this for use internally? or to ease the job for the frontend developers? if so, why would they not just do `f.subscribers.includes(userUid)` ? the way its implemented here, we would have a variable on all favourites that would change depending on who is requesting the object.. or do you foresee 1000++ user subscribing to a object?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java [hunk 3]
```diff
@@ -662,6 +663,37 @@ public void setAsFavorite( @PathVariable( "uid" ) String pvUid, HttpServletReque
         webMessageService.send( WebMessageUtils.ok( message ), response, request );
     }
 
+    @RequestMapping( value = "/{uid}/subscriber", method = RequestMethod.POST )
```
Comment (mortenoh) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: Should that be /subscribe ?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java [hunk 4]
```diff
@@ -822,6 +854,37 @@ public void removeAsFavorite( @PathVariable( "uid" ) String pvUid, HttpServletRe
         webMessageService.send( WebMessageUtils.ok( message ), response, request );
     }
 
+    @RequestMapping( value = "/{uid}/subscriber", method = RequestMethod.DELETE )
```
Comment (mortenoh) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: Should that be /subscribe ?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java [hunk 5]
```diff
@@ -822,6 +854,37 @@ public void removeAsFavorite( @PathVariable( "uid" ) String pvUid, HttpServletRe
...
+
+        if ( entity.isEmpty() )
+        {
+            throw new WebMessageException( WebMessageUtils.notFound( getEntityClass(), pvUid ) );
+        }
+
+        SubscribableObject object = entity.get( 0 );
+        User user = currentUserService.getCurrentUser();
+
+        if ( user == null )
```
Comment (mortenoh) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: This is not really necessary, you can't access the web api without a active security context

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java (hunk 2 above)
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java]: Yes it is to make it easier for consumers / apps I think. We have several fields which change per user e.g. access so I think this is fine.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java (hunk 1 above)
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseAnalyticalObject.java]: I think its okay.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: I think using nouns are better in REST APIs.

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java (hunk 5 above)
Comment (larshelge) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: Okay lets remove that, I have updated the favorites code to reflect it.

Comment (adrianq) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AbstractCrudController.java]: Ok, I have just removed it from the subscribe/unsubscribe methods as well. I also updated the branch with the latest from master...

**Pull request merged**